src/Data/RadixNTree/Word8/Strict.hs

```diff
@@ -0,0 +1,5643 @@
+          case f Nothing of
+            Nothing -> t
+            Just a  -> singleton_ step w s a
+
+
+
+{-# INLINE shape0 #-}
+shape0 :: (RadixTree a -> RadixTree a) -> Feed -> RadixTree a -> RadixTree a
+shape0 f (Feed feed) = \t0@(RadixTree mx t) ->
+  feed $ \step s ->
+    case step s of
+      More w z -> RadixTree mx $ shape_ f step w z t
+      Done     -> f t0
+
+{-# INLINE shape1 #-}
+shape1 :: (RadixTree a -> RadixTree a) -> Feed1 -> Radix1Tree a -> Radix1Tree a
+shape1 f (Feed1 w feed) =
+  feed $ \step -> shape_ f step w
+
+{-# INLINE shape_ #-}
+shape_
+  :: (RadixTree a -> RadixTree a)
+  -> (x -> Step Word8 x) -> Word8 -> x -> Radix1Tree a -> Radix1Tree a
+shape_ f step = go
+  where
+    go w s t =
+      case t of
+        Bin p l r
+          | beyond p w -> let !(RadixTree my dy) = f (RadixTree Nothing Nil)
+                          in case retip (fromStep step w s) my dy of
+                               Nil -> t
+                               dz  -> join
+                                        w dz
+                                        p t
+
+          | w < p      -> rebinL p (go w s l) r
+          | otherwise  -> rebinR p l (go w s r)
+
+        Tip arr mx dx -> goarr w s 0
+          where
+            goarr v z n
+              | v == indexByteArray arr n =
+                  let n' = n + 1
+                  in if n' >= sizeofByteArray arr
+                       then case step z of
+                              More u z' -> retip arr mx (go u z' dx)
+                              Done      -> let !(RadixTree my dy) = f (RadixTree mx dx)
+                                           in retip arr my dy
+
+                       else case step z of
+                              More u z' -> goarr u z' n'
+                              Done      ->
+                                let !(# !brr, !crr #) = splitByteArray 0 n' arr
+
+                                    !(RadixTree my dy) = f (RadixTree Nothing (Tip crr mx dx))
+
+                                in retip brr my dy
+
+              | otherwise =
+                  let !(RadixTree my dy) = f (RadixTree Nothing Nil)
+                  in case retip (fromStep step v z) my dy of
+                       Nil -> t
+                       dz  ->
+                         if n == 0
+                           then join
+                                  (indexByteArray arr 0) (Tip arr mx dx)
+                                  v dz
+
+                           else let !(# !brr, !crr #) = splitByteArray 0 n arr
+                                in Tip brr Nothing $
+                                     join
+                                       (indexByteArray crr 0) (Tip crr mx dx)
+                                       v dz
+
+        Nil       ->
+          let !(RadixTree my dy) = f (RadixTree Nothing Nil)
+          in retip (fromStep step w s) my dy
+
+
+
+-- | Result of a tree split.
+data Split l r = Split !(RadixTree l) !(RadixTree r)
```

Comment:
It seems this is just a strict pair from the `strict` package. There's also strict `Maybe` in there so it would make senso switch.

This echoes another suggestion about unifying function types between lazy and strict versions of the API. Even if no unification is to take place, returning strict tuple from `strict` package seems better thar returning custom type isomorphic to it - the strict tuple has helpers to work with it whereas types like `Split` don't.

Reply:
Those helper functions will be pretty useless when it turns out your code looks like

```
let !(Radix.Split l r) = Radix.split _ _ a
in _
```

At the end of the day this is just the return type for a bunch of functions in this library and if I want to unpack or unbang something later on, I want to have that power. Also means we don't have to depend on three extra libraries.

---

There are broader points here of Haskell suffering from dependency hells and people shunning custom types, but this isn't a good time to rant about it.